compiler+runtime/src/cpp/jank/analyze/processor.cpp

```diff
@@ -2701,69 +2700,83 @@ namespace jank::analyze
+                ->add_usage(read::parse::reparse_nth(item, 2));
             }
-
-            auto const sym(runtime::expect_object<runtime::obj::symbol>(sym_obj));
-            if(!sym->get_namespace().empty())
+            auto const catch_sym(runtime::expect_object<runtime::obj::symbol>(catch_sym_form));
+            if(!catch_sym->get_namespace().empty())
             {
-              return error::analyze_invalid_try("The symbol after 'catch' must be unqualified.",
-                                                object_source(sym_obj),
-                                                latest_expansion(macro_expansions));
+              return error::analyze_invalid_try(
+                "The binding symbol in 'catch' must be unqualified.",
+                object_source(item),
+                latest_expansion(macro_expansions));
             }
-
-            catch_frame->locals.emplace(sym, local_binding{ sym, sym->name, none, catch_frame });
+            auto const catch_type_ref(static_ref_cast<expr::cpp_type>(catch_type.expect_ok()));
+
+            auto catch_frame(
+              jtl::make_ref<local_frame>(local_frame::frame_type::catch_, current_frame));
+            catch_frame->locals.emplace(catch_sym,
+                                        local_binding{ catch_sym,
+                                                       catch_sym->name,
+                                                       none,
+                                                       catch_frame,
+                                                       true,
+                                                       false,
+                                                       false,
+                                                       catch_type_ref->type });
 
             /* Now we just turn the body into a do block and have the do analyzer handle the rest. */
-            auto const do_list(
-              catch_list->data.rest().rest().conj(make_box<runtime::obj::symbol>("do")));
+            auto const do_list(catch_it.rest().conj(make_box<runtime::obj::symbol>("do")));
             auto const do_res(analyze(make_box(do_list), catch_frame, position, fn_ctx, true));
             if(do_res.is_err())
             {
               return do_res.expect_err();
             }
-
-            /* TODO: Read this from the catch form. */
-            static auto const object_ref_type{ cpp_util::resolve_literal_type(
-                                                 "jank::runtime::oref<jank::runtime::object>")
-                                                 .expect_ok() };
-
-            ret->catch_body = expr::catch_{ sym,
-                                            object_ref_type,
-                                            static_ref_cast<expr::do_>(do_res.expect_ok()) };
+            do_res.expect_ok()->frame = catch_frame;
+            ret->catch_bodies.emplace_back(
+              expr::catch_{ catch_sym,
```

Comment:
If you get rid of the option, you can also get rid of the `expr::catch_{` here and just pass these values directly.